### PR TITLE
[posix] Introduce nslookup table

### PR DESCRIPTION
**Pull request by jkeljo**
Query by `address`, get `hostname`. Query by `hostname`, get `address`(es):

```
osquery> select * from nslookup where address = '8.8.8.8';
+---------+--------------------------------+
| address | hostname                       |
+---------+--------------------------------+
| 8.8.8.8 | google-public-dns-a.google.com |
+---------+--------------------------------+
osquery> select * from nslookup where hostname = 'google-public-dns-a.google.com';
+----------------------+--------------------------------+
| address              | hostname                       |
+----------------------+--------------------------------+
| 0:0:2001:4860:4860:: | google-public-dns-a.google.com |
| 8.8.8.8              | google-public-dns-a.google.com |
+----------------------+--------------------------------+
```

### CODE REVIEW

**Comment by jkeljo** [comment]
There is something **really** slow about joining this with another table; e.g.:
`select * from interface_addresses inner join nslookup using (address) where interface like 'en%';`
on my Mac calls `genLookup` four times, each of which completes super fast, but the overall query takes 5 seconds to come back. Instruments doesn't show much going on in the CPU department, so I'm not sure what the deal is. 

I have to leave for the day but I wanted to get this up and get some feedback on it, also see if anyone  has any insight on the slow join issue before I dig in further.

**Comment by osqueryer** [comment]
:-1: The commit 6098a2002def10ac231c9b23f478b037cd230525 (Job results: [1723](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1723/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 6098a2002def10ac231c9b23f478b037cd230525 (Job results: [1725](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1725/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 6098a2002def10ac231c9b23f478b037cd230525 (Job results: [4225](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4225/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 6098a2002def10ac231c9b23f478b037cd230525 (Job results: [4227](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4227/)) failed one or more tests (**Windows**).

**Comment by obelisk** [comment]
You can try running it with --planner and see if that is illuminating.

File: osquery/tables/networking/posix/nslookup.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+QueryData genLookup(QueryContext& context) {
+  QueryData results;
+
+  if (context.constraints.count("address") > 0 &&
+      context.constraints.at("address").exists(EQUALS)) {
+    for (const auto& ip_str : context.constraints.at("address").getAll(EQUALS)) {
+      struct sockaddr_in ip4addr;
+      struct sockaddr_in6 ip6addr;
+      const struct sockaddr* addr =
```
Comment (allanliu) [inline osquery/tables/networking/posix/nslookup.cpp]:
```c++ 
auto addr = reinterpret_cast<const struct sockaddr *>(&ip4addr);
``` 

Using auto is generally prefered.

File: osquery/tables/networking/posix/nslookup.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+#include <osquery/tables.h>
+
+namespace osquery {
+namespace tables {
+
+QueryData genLookup(QueryContext& context) {
+  QueryData results;
+
+  if (context.constraints.count("address") > 0 &&
+      context.constraints.at("address").exists(EQUALS)) {
```
Comment (allanliu) [inline osquery/tables/networking/posix/nslookup.cpp]: I think you can make a minor optimization here by avoiding some extra lookups...
```c++
auto addrs = context.constraints.find("address");
if (addrs != context.contraints.end()  &&
    addrs->second.exists(EQUALS)) {
    ...
    for (const auto& ip_str : addrs.getAll(EQUALS)) {
      ...
    }
```

File: osquery/tables/networking/posix/nslookup.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+      Row r;
+      r["address"] = ip_str;
+      r["hostname"] = hostname;
+      results.push_back(r);
+    }
+  }
+
+  if (context.constraints.count("hostname") > 0 &&
+      context.constraints.at("hostname").exists(EQUALS)) {
```
Comment (allanliu) [inline osquery/tables/networking/posix/nslookup.cpp]: Same as above.

File: osquery/tables/networking/posix/nslookup.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        &hints,
+        &addr0) != 0) {
+          continue;
+      }
+
+      struct addrinfo* addr = addr0;
+      while (addr != nullptr) {
+        char ip_address_str[INET6_ADDRSTRLEN];
+        struct sockaddr_in* sockaddr =
+          reinterpret_cast<struct sockaddr_in *>(addr->ai_addr);
```
Comment (allanliu) [inline osquery/tables/networking/posix/nslookup.cpp]: Nit: use auto, same as above.

**Comment by theopolis** [comment]
ping @jkeljo!

**Comment by Smjert** [comment]
I'm going to close this since there has not been any further interaction after the first review and the subsequent ping.

**Pull request closed without merging**